rohmu/notifier/http.py

```diff
@@ -0,0 +1,132 @@
+            self._url,
+            {
+                "op": Operation.UPLOAD.value,
+                "key": key,
+                "size": size,
+                "last_modified": datetime.now(),
+            },
+        )
+
+    def object_deleted(self, key: str) -> None:
+        try:
+            self._session.post(self._url, {"op": Operation.DELETE.value, "key": key})
+        except RequestException as e:
+            LOG.warning(
+                "[HTTPNotifier] POST request to %r failed with %r, dropping event",
+                self._url,
+                str(e),
+            )
+
+    def tree_deleted(self, key: str) -> None:
+        try:
+            self._session.post(self._url, {"op": Operation.DELETE_TREE.value, "key": key})
+        except RequestException as e:
+            LOG.warning(
+                "[HTTPNotifier] POST request to %r failed with %r, dropping event",
+                self._url,
+                str(e),
+            )
+
+
+@dataclass
+class _HTTPNotifyJob:
+    url: str
+    data: Dict[str, Any]
+
+
+def _background_http_request(queue: "SimpleQueue[_HTTPNotifyJob]") -> None:
+    with _get_requests_session() as session:
+        while True:
+            notification = queue.get()
+
+            try:
+                session.post(notification.url, notification.data)
+            except RequestException as e:
+                # The session is configured to do automatic retries, and that should handle most
+                # transients errors. Situations that takes longer than the configured retries will
+                # cause events to be dropped because we can't differentiate it from a non-transient
+                # errors.
+                #
+                # For non-transient errors the thread must kept running to consume from the queue,
+                # because there is no mechanism to stop producing to it. Hopefully the failure is
+                # fast enough that the queue won't grow faster than this threads consumes from it.
+                LOG.warning(
+                    "[BackgroundHTTPNotifier] POST request to %r failed with %r, dropping event",
+                    notification.url,
+                    str(e),
+                )
+
+
+class BackgroundHTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._url = url
+        self._queue: "SimpleQueue[_HTTPNotifyJob]" = SimpleQueue()
+        self._thread = threading.Thread(
+            target=_background_http_request,
+            args=(self._queue,),
+            name="BackgroundHTTPNotifierThread",
+            daemon=True,
+        )
+        self._thread.start()
+
+    def object_created(self, key: str, size: Optional[int]) -> None:
+        self._queue.put(
+            _HTTPNotifyJob(
+                self._url,
+                {
+                    "op": Operation.UPLOAD.value,
+                    "key": key,
+                    "size": size,
+                    "last_modified": datetime.now(),
```

Comment:
This should probably be `utcnow`, unless we want interesting bugs during summer time :)

Reply:
changed to a time-aware datatime with UTC.